codex/sales.nim

```diff
@@ -120,27 +129,238 @@ proc load*(sales: Sales) {.async.} =
       slot.request.id,
       slot.slotIndex,
       some slot.request)
+
+    agent.context.onCleanUp = proc {.async.} = await sales.remove(agent)
+
     agent.start(SaleUnknown())
     sales.agents.add agent
 
-proc start*(sales: Sales) {.async.} =
-  doAssert sales.subscription.isNone, "Sales already started"
+proc subscribeRequested(sales: Sales) {.async.} =
+  let context = sales.context
+  let market = context.market
+  let subs = sales.subscriptions
+
+  if subs.requested.isSome:
+    return
 
-  proc onRequest(requestId: RequestId, ask: StorageAsk) {.gcsafe, upraises:[].} =
-    sales.handleRequest(requestId, ask)
+  proc onRequestEvent(requestId: RequestId,
+                      ask: StorageAsk,
+                      expiry: UInt256) {.gcsafe, upraises:[].} =
+    try:
+      let reservations = context.reservations
+      let slotQueue = context.slotQueue
+      # Match availability before pushing. If availabilities weren't matched,
+      # every request in the network would get added to the slot queue.
+      # However, matching availabilities requires the subscription callback to
+      # be async, which has been avoided on many occasions, so we are using
+      # `waitFor`.
+      if availability =? waitFor reservations.find(ask.slotSize,
+                                                   ask.duration,
+                                                   ask.pricePerSlot,
+                                                   ask.collateral,
+                                                   used = false):
```

Comment:
> Shouldn't this happen inside the queue?

It's a good idea for it to be added as a check for pushing to the queue, if we still need that check (more on this in the main comments)

> I get that we don't want every request to be inserted into the queue, but you are asserting here something (eq. we have space for the request) which then might change because some other request will take over the priority and might fill the availability inside the queue.

Yes, true. There are two availabilities checks. One is done before pushing to the queue. The second is done once the item is popped from the queue before processing. The first check is, as you mentioned, to prevent putting every request onto the queue. The second check is done, as you also mentioned, to check that there is still availability which may have changed since initially added. I thought I had put the first AND second checks in the [slot queue design](https://github.com/codex-storage/codex-research/blob/master/design/sales.md#slot-queue), but after re-reading, the first check was not added. It might be better without the first check as you've pointed out in your main comment.

> I would like for this to happen inside the state machine as a first step.

The availability check is already happening [inside of the state machine as a first step (`states/preparing.nim`)](https://github.com/codex-storage/nim-codex/pull/455/files#diff-cc836926ac647abfeb4b80f354bd11f7e8495fdfbff17ab9fdc5d06afa1b95c0R53-R64)

> Then we don't need the waitFor, which I'd really like to avoid.

Respectfully, I disagree that this is a good reason to move the availabilities check.